### PR TITLE
RFC3: document message part framing, and message framing on UNIX stream socket

### PR DESCRIPTION
**Pull request by garlick**
As noted in #344, Flux has taken over the framing of message parts from ZeroMQ, but we neglected to document that.  Also, we never documented how full messages are framed on a UNIX stream socket by the `local` connector.

This fills in those missing pieces.

RFC 3 was written a while ago and when I read it with my "write a new message encoder/transport" hat on, it feels it doesn't have the right structure.  I'd like to improve that, but for now just wanted to put in the missing stuff.

### CODE REVIEW

**Comment by vsoch** [comment]
I was going to try playing around with this today (and got totally consumed with a personal project!) I will definitely play around with this soon.

**Comment by garlick** [comment]
Cool! I was trying to remember if there is anything I missed.   Like it seems like we had a hack for the local connector where a message was sent from server to client on the initial connection to provide the instance owner userid or something, but I'm not seeing that in a message trace at least, e.g.
```
garlick@picl0:~$ FLUX_HANDLE_TRACE=1 flux getattr rank
--------------------------------------
> 0.00000
> attr.get
> flags=topic,payload,route userid=unknown rolemask=none nodeid=any matchtag=1
> {"name":"rank"}
--------------------------------------
< 0.00130
< attr.get
< flags=topic,payload,route userid=500 rolemask=owner errnum=0 matchtag=1
< {"value":"0","flags":1}
0
```
Deeper spelunking required than i had time for today...

**Comment by garlick** [comment]
> it seems like we had a hack for the local connector where a message was sent from server to client on the initial connection to provide the instance owner userid or something

For the record, I confirmed that is ancient history, no longer a thing.

File: spec_3.rst [hunk 1]
```diff
@@ -215,29 +215,54 @@ established for common payload types:
 5. JSON payloads SHALL include a terminating NULL character.
 
 
-General Message Format
-~~~~~~~~~~~~~~~~~~~~~~
+Message Structure
+~~~~~~~~~~~~~~~~~
```
Comment (chu11) [inline spec_3.rst:219]: perhaps an example or two might be a good idea here?

File: spec_3.rst [hunk 2]
```diff
@@ -341,6 +341,19 @@ Flux messages are specified by the following modified ABNF grammar [#f2]_
    ; unused 4-byte field
    unused          = %x00.00.00.00
 
+
+Message Framing
+~~~~~~~~~~~~~~~
+
+When Flux uses ZeroMQ as transport (``tcp``, ``ipc``, and ``shmem``
+connectors), ZeroMQ handles the message framing.
```
Comment (chu11) [inline spec_3.rst:349]: "as transport" doesn't sound right.  "as a transport"?  "for transport"?

File: spec_3.rst (hunk 1 above)
Comment (garlick) [reply spec_3.rst:219]: Agreed.  Just pushed a couple of examples.  Helpful?

Comment (garlick) [reply spec_3.rst:219]: oops force pushed again to fix duplicate captions in the image, and add some circles and arrows to further explain.

Comment (chu11) [reply spec_3.rst:219]: should the blue column prefix with `0x`?  Took me awhile to realize it wasn't decimal :-)

Comment (garlick) [reply spec_3.rst:219]: I was tryiing to be consistent throughout and 0x would be kind of ugly - maybe if I just incorporate the footnote into the caption it'll help.  I'll push that - hang on...

**Review by chu11 — APPROVED** [review]
LGTM

**Comment by garlick** [comment]
Thanks! I'll set MWP.

**Pull request merged**